File: algorithms/merge_sort/merge_sort.js
var mergeSort = function(array) {
  if (array.length <= 1) {
    return array;
  }

  var middle = Math.floor(array.length / 2);
  var left = array.slice(0, middle);
  var right = array.slice(middle);

  return merge(mergeSort(left), mergeSort(right));
};

var merge = function(left, right) {
  var result = [];

  while(left.length > 0 && right.length > 0) {
    if (left[0] < right[0]) {
      result.push(left.shift());
    } else {
      result.push(right.shift());
    }
  }

  return result.concat(left).concat(right);
};

console.log(mergeSort([5, 3, 2, 19, 15, 25, 7, 9]));

File: algorithms/merge_sort/merge_sort.rb
def merge_sort(array)
  return array if array.length <= 1
  midpoint = array.length / 2
  left_half = array.take(midpoint)
  right_half = array.drop(midpoint)
  merge(merge_sort(left_half), merge_sort(right_half))
end

def merge(left_half, right_half)
  result = []
  until left_half.empty? || right_half.empty?
    result << ((left_half.first < right_half.first) ?
      left_half.shift : right_half.shift)
  end
  result + left_half + right_half
end

if __FILE__ == $PROGRAM_NAME
  p merge_sort([5, 3, 2, 19, 15, 25, 7, 9])
end
